Feature/publish occupancy grid from tango 3d reconstruction

PerrineAguiar opened the pull request:
This PR adds the feature to publish an occupancy grid from Tango 3d reconstruction.
It uses the latest API of Tango 3d reconstruction which is not yet officially released.

Note that when updating the Tango 3d reconstruction library, some changes were required as explained in the [Hopak release note](https://developers.google.com/tango/release-notes#may_2017_hopak).

EDIT
The API used to create an occupancy grid should be available in the next release (I release). This PR will be merged as soon as it is officially released.

Fixes #26.

mcopejans commented on tango_ros_common/tango_ros_native/include/tango_ros_native/tango_ros_conversions_helper.h:
```diff
@@ -27,6 +28,7 @@
 
 namespace tango_ros_conversions_helper {
 const int NUMBER_OF_FIELDS_IN_POINT_CLOUD = 4;
+const float OCCUPANCY_GRID_RESOLUTION = 0.05;
```
What is the unit? Is it in meters?

mcopejans commented on tango_ros_common/tango_ros_native/include/tango_ros_native/tango_ros_conversions_helper.h:
```diff
@@ -112,10 +114,22 @@ void toTango3DR_Pose(const TangoPoseData& tango_pose_data, Tango3DR_Pose* t3dr_p
 // @param mesh_marker, the output visualization_msgs::Marker.
 void toMeshMarker(const Tango3DR_GridIndex& grid_index,
-                  Tango3DR_Mesh* tango_mesh,
+                  const Tango3DR_Mesh& tango_mesh,
```
:+1:

mcopejans commented on tango_ros_common/tango_ros_native/src/tango_ros_conversions_helper.cpp:
```diff
@@ -232,26 +234,62 @@ void toMeshMarker(const Tango3DR_GridIndex& grid_index,
+      int value = static_cast<int>(image_grid.data[j + (image_grid.height - i - 1) * image_grid.width]);
+      if (value == 1) {
+        // The cell is occupied.
+        occupancy_grid->data.push_back(100);
```
Some performance gain can be achieved by reserving memory for the data vector: `occupancy_grid->data.reserve(value)` with value equal to `image_grid.height * image_grid.width`. Is this suggestion feasible in this case?

mcopejans commented on tango_ros_common/tango_ros_native/src/tango_ros_conversions_helper.cpp:
```diff
@@ -232,26 +234,62 @@ void toMeshMarker(const Tango3DR_GridIndex& grid_index,
+        occupancy_grid->data.push_back(100);
+      } else if (value == 255) {
+        // The cell is free.
+        occupancy_grid->data.push_back(0);
```
Using meaningful names for constants instead of documentation for `0, -1, 100` would also work.
constexpr int FREE = 0 (same for UNKNOWN, OCCUPIED)

mcopejans commented on tango_ros_common/tango_ros_native/src/tango_ros_node.cpp:
```diff
@@ -418,11 +421,12 @@ TangoErrorType TangoRosNode::TangoSetupConfig() {
   const char* resolution = "resolution";
-  result = Tango3DR_Config_setDouble(t3dr_config, resolution, 0.05);
+  result = Tango3DR_Config_setDouble(t3dr_config, resolution,
+                                     tango_ros_conversions_helper::OCCUPANCY_GRID_RESOLUTION);
```
Is this a value that a user might want to change depending on the scale of his application?

PerrineAguiar replied:
Actually yes, we should make it a ros parameter.

mcopejans commented on tango_ros_common/tango_ros_native/src/tango_ros_node.cpp:
```diff
@@ -974,67 +983,82 @@ void TangoRosNode::PublishMeshMarker() {
+      // Publish Tango mesh as visualization marker.
+      if (mesh_marker_publisher_.getNumSubscribers() > 0) {
+        visualization_msgs::MarkerArray mesh_marker_array;
+        for (size_t i = 0; i < t3dr_updated_indices.num_indices; ++i) {
```
Do you think it is useful to make a helper function `toMeshMarkerArray` instead of having a for loop here?

smits replied:
I think we should try to keep the extract functions out of the conversion functions, I'm not sure if that would work with your suggestion.

mcopejans commented on third_party/tango_api/tango_3d_reconstruction/include/tango_3d_reconstruction/tango_3d_reconstruction_api.h:
```diff
@@ -47,10 +47,10 @@ typedef enum {
+  TANGO_3DR_HAL_PIXEL_FORMAT_RGBA_8888 = 1,         ///< RGBA 8888
+  TANGO_3DR_HAL_PIXEL_FORMAT_RGB_888 = 3,           ///< RGB 888
+  TANGO_3DR_HAL_PIXEL_FORMAT_YCrCb_420_SP = 0x11,   ///< NV21
+  TANGO_3DR_HAL_PIXEL_FORMAT_DEPTH16 = 0x44363159,  ///< DEPTH16
```
So these are changes from master? Would it make sense to have this header as a submodule from [tango examples c](https://github.com/googlesamples/tango-examples-c/blob/master/tango_3d_reconstruction/include/tango_3d_reconstruction_api.h) github?

PerrineAguiar replied:
Yes. However the changes in [tango examples c ](https://github.com/googlesamples/tango-examples-c/blob/master/tango_3d_reconstruction/include/tango_3d_reconstruction_api.h) does not seem up to date. I don't see the changes mentioned in the [Hopak release note](https://developers.google.com/tango/release-notes#may_2017_hopak). Same if I download the inlcuded and libs from [here](https://developers.google.com/tango/downloads).

mcopejans replied:
And of hou switch to release-hopa branch? Are they correct here? https://github.com/googlesamples/tango-examples-c/blob/release-hopka/tango_3d_reconstruction/include/tango_3d_reconstruction_api.h

PerrineAguiar replied:
After some more checking, I think that `tango-examples-c` (master or hopak branch) is up to date with the release note, I was checking the wrong function.
Some other changes are not in Hopak yet, I suppose they will be in the next release. Among these changes, this PR depends on the new `Tango3DR_extractFullFloorplanImage` API. Therefore, I think we will have to wait for the next release to merge this PR.

mcopejans replied:
Where did this header come from?

PerrineAguiar commented:
@mcopejans Thanks for your review. 
The new API should be in the next release. I'll wait for it to be officially released to add the header as submodule and merge this PR.

smits commented on tango_ros_common/tango_ros_native/src/tango_ros_node.cpp:
```diff
@@ -762,17 +768,17 @@ void TangoRosNode::StartPublishing() {
-      if (!tango_data_available_ || !publish_pose_on_tf_
-          || !publish_pose_on_topic_) {
+      if (!tango_data_available_ || (!publish_pose_on_tf_
+          && !publish_pose_on_topic_)) {
```
I guess this will need to be cleaned up when merged with master as the publish_pose_on_topic will be gone?

smits commented on tango_ros_common/tango_ros_native/src/tango_ros_node.cpp:
```diff
@@ -974,67 +983,82 @@ void TangoRosNode::PublishMeshMarker() {
+          Tango3DR_Vector2 origin;
+          Tango3DR_ImageBuffer image_grid;
+          result = Tango3DR_extractFullFloorplanImage(
+              t3dr_context_, TANGO_3DR_LAYER_OBSTACLES, &origin, &image_grid);
```
Would it be an idea to make the layer configurable? Or even support extracting different layers? @juergensturm what do you think.  Could be in a follow up though.

juergensturm replied:
Yes, in general this is a good idea. But at the moment, I don't see a strong use case for this, so we can keep it as a low prio feature request for now.

PerrineAguiar commented on tango_ros_common/tango_ros_native/include/tango_ros_native/tango_ros_conversions_helper.h:
```diff
@@ -27,6 +28,10 @@
 
 namespace tango_ros_conversions_helper {
 const int NUMBER_OF_FIELDS_IN_POINT_CLOUD = 4;
+const float OCCUPANCY_GRID_RESOLUTION = 0.05; // in meter.
```
Done in #304.

PerrineAguiar commented:
@smits @mcopejans Thanks for your reviews, I addressed your comments, ptal.
The new tango 3d reconstruction API needed for this PR has been released (Ikariotikos release), so this PR can be merged.
As suggested by @mcopejans I added the `tango-examples-c` repository as a submodule and linked the tango libs and headers from it.

mcopejans commented:
Lgtm!
